### PR TITLE
Fix and import checkout registration from spree_auth_devise

### PR DESCRIPTION
**Pull request by mbj**
Fix email validation on registration
    
    * The registration step used to bypass email validation
    * The code was "buried" in even worse specified `spree_auth_devise`
      hence was lifted to spree
    * We cannot yet lift the views as they reference routes provided by
      devise hence we only provide a stup here to ease the incremental
      change needed.

### CODE REVIEW

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 1]
```diff
@@ -16,11 +16,99 @@
...
+      ensure_sufficient_stock_lines: nil,
+      set_state_if_present:          nil
+    )
+  end
+
+  shared_examples 'renders #registration' do
+    it 'assigns a new user model' do
+      do_request
+
+      expect(assigns[:user].instance_of?(Spree::User)).to be(true)
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: This can be written as:

```ruby
expect(assigns[:user]).to be_instance_of?(Spree::User)
```

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -64,7 +69,50 @@ def update
...
+    # Render registration
+    #
+    # @return [undefined]
+    def render_registration
+      @user = User.new
+      render('registration')
+    end
+
+    def check_registration
+      return if try_spree_current_user || current_order.email
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Is it possible for `current_order.email` to be `''` thus causing this boolean to pass when maybe it should not? Maybe this should use `current_order.email?` to ensure that the email is present.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: > Is it possible for current_order.email to be '' thus causing this boolean to pass when maybe it should not? Maybe this should use current_order.email? to ensure that the email is present

In theory an invalid email address (such as the empty string) should not reach `#email` but I'll double check this. good catch.

Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: `#email` is prevented to ever evaluate to the empty string by prior validations.

**Comment by mbj** [comment]
@dkubb This is ready for the next pass.

**Comment by dkubb** [comment]
@mbj I assume when this is merged you will remove the similar code from `spree_auth_devise`?

**Comment by mbj** [comment]
> @mbj I assume when this is merged you will remove the similar code from spree_auth_devise?

Yes, that PR already exists and is green: https://github.com/MountainRoseHerbs/spree_auth_devise/pull/24

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 3]
```diff
@@ -64,7 +69,50 @@ def update
...
+    def update_registration
+      email = params.fetch(:order).fetch(:email)
+
+      if cart_order.update_attributes(email: email)
+        redirect_to(checkout_path)
+      else
+        flash[:registration_error] = cart_order
+          .errors
+          .full_messages
+          .join(MESSAGE_DELIMITER)
```
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Not a blocker, but the `object.errors.full_messages.join(', ')` pattern appears in 7 places within `spree` and 1-2 places in `mrh-spree`.

File: frontend/app/controllers/spree/checkout_controller.rb (hunk 2 above)
Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: Would it still be semantically superior to use a simpler boolean method like `#email?` here instead of testing the truthy-ness of a non-boolean method?

**Comment by dkubb** [comment]
The first commit in this series has a spelling mistake in the git commit summary. I can't tell if you meant `step` or `setup`.

**Comment by dkubb** [comment]
@mbj I realize this is simply moving code, but would it be better (in the future) to extract the registration stuff into it's own controller? We would be able to then use a more standard pairing of `edit` and `update` actions instead of custom named actions.

**Comment by dkubb** [comment]
@mbj there's a spelling mistake in your second git commit, eg: (missing the "s" in address)

```
Avoid to not require email in addres state
```

Also I'm not sure about this grammar, it's unclear to me what you meant to say.

**Comment by dkubb** [comment]
@mbj if the address state is never reached in the db, does that mean we can remove the state from the `order_state` type?

**Comment by mbj** [comment]
> @mbj I realize this is simply moving code

Moving and fixing, the fact the email was written outside the model validations is the reason for all this.

> but would it be better (in the future) to extract the registration stuff into it's own controller? We would be able to then use a more standard pairing of edit and update actions instead of custom named actions.

There is already a `UsersController` I suspect this is the right place, for now I intentionally only moved the fixed code and not extracted it to cut the diff.

I think we should move over all code from `spree_auth_devise` before we do more architectural changes.

**Comment by mbj** [comment]
> @mbj if the address state is never reached in the db, does that mean we can remove the state from the order_state type?

Yes, but that is better done in a separate iteration.

**Comment by mbj** [comment]
> @mbj there's a spelling mistake in your second git commit, eg: (missing the "s" in address)

fixed. (also the grammar)

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 4]
```diff
@@ -438,17 +519,6 @@ def do_request
...
-          context 'and order user is not present' do
-            let(:user) { nil }
-
-            it 'does not attempt to persist user' do
-              # We do not have a user to place a negative message expectation on
-              # Hence for covering the semantics its sufficient to assert the code
-              # does not crash in this setup.
-              do_request
-            end
-          end
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: Is this context removed because the other two contexts added later make it redundant, or was it removed because it began failing? If it's the former, I might suggest removing it in it's own commit, with a message saying it's being removed due to redundancy.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb [hunk 5]
```diff
@@ -479,6 +549,33 @@ def do_request
...
+
+          expect(order).to have_received(:next)
+        end
+      end
+
+      context 'and order has no email set' do
+        it 'should redirect to registration' do
+          do_request
+
+          expect(session['spree_user_return_to']).to eql(spree.checkout_state_path('address'))
```
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: Could `'address'` be replaced with `order_state`?

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 4 above)
Comment (dkubb) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: From my basic testing it appears this is made redundant, so I moved it into it's own commit.

File: frontend/spec/controllers/spree/checkout_controller_spec.rb (hunk 5 above)
Comment (mbj) [inline frontend/spec/controllers/spree/checkout_controller_spec.rb]: yeah

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 6]
```diff
@@ -109,7 +109,7 @@ def render_registration
     end
 
     def check_registration
-      return if try_spree_current_user || current_order.email
+      return if try_spree_current_user || current_order.email?
```
Comment (mbj) [inline frontend/app/controllers/spree/checkout_controller.rb]: Do we actually need this predicate? (I think its logically the right choice to use it anyway) but afaik empty string email addresses could never reach here.

Comment (dkubb) [inline frontend/app/controllers/spree/checkout_controller.rb]: It is not strictly required, but semantically I feel it's simpler than doing a full test of the truthy-ness of the attribute value.
